### PR TITLE
ISSUE-352: Durability fails to deserialize rules containing condition…

### PR DESCRIPTION
**Pull request by EthanEChristian**
…s with an empty constraint

### CODE REVIEW

**Comment by EthanEChristian** [comment]
Issue 352

**Comment by EthanEChristian** [comment]
@mrrodriguez @WilliamParker

File: src/test/clojure/clara/durability_rules.clj [hunk 1]
```diff
@@ -29,7 +29,10 @@
 
 (defrule find-wind-speeds-without-temp
   "Rule using NegationNode"
-  [?w <- WindSpeed (= ?loc location)]
+  [?w <- WindSpeed
+   ;; Empty constraint to test the serializing EmptyList, see Issue 352 for more information
+   ()
```
Comment (EthanEChristian) [inline src/test/clojure/clara/durability_rules.clj]: I'm not sure that its right to accept this, however removing it as a valid constraint would probably be seen as non-passive

File: src/main/clojure/clara/rules/durability/fressian.clj [hunk 2]
```diff
@@ -235,6 +235,16 @@
                 (read [_ rdr tag component-count]
                   (read-with-meta rdr #(apply list %))))}}
 
+   "clj/emptylist"
+   {:class clojure.lang.PersistentList$EmptyList
```
Comment (mrrodriguez) [inline src/main/clojure/clara/rules/durability/fressian.clj]: This is a sneaky case. The case for `clojure.lang.ASeq` is meant to catch the Clojure collection types not otherwise specified here, but  `clojure.lang.PersistentList$EmptyList` is a special inner class that apparently only implements the interfaces, which differs from how the mos to the other Clojure data structures are implemented.

However, it is concerning. Classes like this are really "implementation details" of Clojure. I don't know that there is a good way around this though, other than to keep it consistent with Clara's supported versions of Clojure.

One, potentially obvious, sounding fix for this sort of issue would be to implement readers and writers against the interfaces instead of these concrete classes. However, this can lead to issues with multiple handlers matching for some objects, which will fail. This is similar to the multiple inheritance sort of issues that happen with protocol extensions. Many concrete collection types will implement many common interfaces. It tends to cause problems quickly. The recommendation I've seen is to only target extending concrete classes instead of interfaces. (Technically the "lookup" mechanism of Fressian is extensible and could be used to implement something similar to multimethod "prefers", but that still gets complicated and brittle.)

The downside, is that we have to be sure to cover all possible types that Clara may be serializing in its own network and working memory structures.

I don't see a better way right now other than doing what you did here and just adding the classes. This made me look around in Clojure to recall if there are any other special classes used to represent "empty" collections. It doesn't see to be the case for vectors, sets, or maps. Clojure eventually backed off an earlier [approach to using different classes for certain "smaller" collection sizes](https://dev.clojure.org/jira/browse/CLJ-1517) too, so that shouldn't be a problem.

File: src/test/clojure/clara/durability_rules.clj (hunk 1 above)
Comment (mrrodriguez) [reply src/test/clojure/clara/durability_rules.clj]: I think it is a bit odd to have the empty constraint along with something else. However, I agree that it doesn't really make sense to make it start failing now. It seems like something that could be done by people in practice fairly easily or maybe even intentionally for some visual preference.

Also, the empty list is "truthy" in Clojure, so I'd expect it to be evaluated that way.

Comment (EthanEChristian) [reply src/test/clojure/clara/durability_rules.clj]: Yes this is an odd scenario, however i wanted to minimize the change to the rules that we use to test serialization, because one small change can vastly change the tests

File: src/main/clojure/clara/rules/durability/fressian.clj (hunk 2 above)
Comment (WilliamParker) [reply src/main/clojure/clara/rules/durability/fressian.clj]: Yeah, as far as I can tell there is no consistent convention or API in Clojure that would allow us to serialize and deserialize these collections while maintaining their concrete class without special-casing each concrete type.  It is admittedly unfortunate that something like [Java's collection constructor](https://docs.oracle.com/javase/8/docs/api/java/util/ArrayList.html#ArrayList-java.util.Collection-) isn't available for Clojure's data structures.

File: src/test/common/clara/test_simple_rules.cljc [hunk 3]
```diff
@@ -194,3 +194,14 @@
     (is (has-fact? (:cold @side-effect-holder) (->Temperature -10 "MCI") ))
 
     (is (has-fact? (:subzero @side-effect-holder) (->Temperature -10 "MCI")))))
+
+(def-rules-test test-condition-with-empty-constraint
+  {:rules [simple-rule [[[Cold ()]]
```
Comment (WilliamParker) [inline src/test/common/clara/test_simple_rules.cljc]: Can we put the empty list somewhere also where serialization of it will be required even if the empty constraint is filtered out?  Say [Cold () (not= temperature ()]

Comment (EthanEChristian) [reply src/test/common/clara/test_simple_rules.cljc]: Added this to the basic and deserialization test

File: src/main/clojure/clara/rules/durability/fressian.clj [hunk 4]
```diff
@@ -235,6 +235,16 @@
...
+    :writer (reify WriteHandler
+              (write [_ w o]
+                (write-with-meta w "clj/emptylist" o)))
```
Comment (WilliamParker) [inline src/main/clojure/clara/rules/durability/fressian.clj]: Since this type is, by definition, empty, we don't actually need to write out or read in the list itself - we could just write out the metadata if present (or null if absent).  See the implementation of write-with-meta and read-with-meta.

File: src/test/common/clara/test_simple_rules.cljc [hunk 5]
```diff
@@ -194,3 +194,14 @@
     (is (has-fact? (:cold @side-effect-holder) (->Temperature -10 "MCI") ))
 
     (is (has-fact? (:subzero @side-effect-holder) (->Temperature -10 "MCI")))))
+
+(def-rules-test test-condition-with-empty-constraint
+  {:rules [simple-rule [[[Cold () (not= this ())]]
```
Comment (WilliamParker) [inline src/test/common/clara/test_simple_rules.cljc]: This is good, but looking over compile-constraints I think we should have a test where we validate that a rule with only empty constraints, preferably for both the case of 1 empty constraint and 2 empty constraints, works correctly.  The case of no constraints (also a recursion termination case here) is specifically handled with this:

````
(if (empty? exp-seq)
     `(deref ~'?__bindings__)
````

but I'd like to make sure that we're testing the recursion properly both when there are later non-empty constraints and when we bottom out without finding any meaningful constraints.

File: src/test/common/clara/test_simple_rules.cljc [hunk 6]
```diff
@@ -194,3 +194,14 @@
     (is (has-fact? (:cold @side-effect-holder) (->Temperature -10 "MCI") ))
 
     (is (has-fact? (:subzero @side-effect-holder) (->Temperature -10 "MCI")))))
```
Comment (WilliamParker) [inline src/test/common/clara/test_simple_rules.cljc]: This doesn't really seem like a "simple rule" sanity test to me.  I suppose we could put it here, but IMO it fits better in test-dsl since it revolves around edge cases of ways to represent the logical relationships implied by rules, not edge cases of the logical relationships themselves.  Test-dsl is a cross-platform test file as well.

File: src/test/clojure/clara/durability_rules.clj [hunk 7]
```diff
@@ -29,7 +29,12 @@
 
 (defrule find-wind-speeds-without-temp
   "Rule using NegationNode"
-  [?w <- WindSpeed (= ?loc location)]
+  [?w <- WindSpeed
+   ;; Empty constraint and a constraint containing an empty list to test the serializing EmptyList,
```
Comment (WilliamParker) [inline src/test/clojure/clara/durability_rules.clj]: minor grammar: "the serializing" => "serializing the"

**Comment by WilliamParker** [comment]
@EthanEChristian the idea looks good, I left some comments with a few suggestions.

**Comment by EthanEChristian** [comment]
I have address the concerns you mentioned @WilliamParker

File: src/main/clojure/clara/rules/durability/fressian.clj [hunk 8]
```diff
@@ -235,6 +235,23 @@
                 (read [_ rdr tag component-count]
                   (read-with-meta rdr #(apply list %))))}}
 
+   "clj/emptylist"
+   {:class clojure.lang.PersistentList$EmptyList
+    :writer (reify WriteHandler
+              (write [_ w o]
+                (let [m (meta o)]
+                  (do
+                    (.writeTag w "clj/emptylist" 2)
```
Comment (WilliamParker) [inline src/main/clojure/clara/rules/durability/fressian.clj]: Now that there's only one component written out this should be 1, not 2.  See the [Fressian custom handler documentation](https://github.com/clojure/data.fressian/wiki/Creating-custom-handlers); the implementation of this method in [FressianWriter](https://github.com/Datomic/fressian/blob/fressian-0.6.6/src/org/fressian/FressianWriter.java#L252) might be good to glance at as well.

**Comment by WilliamParker** [comment]
@EthanEChristian looks good to me apart from https://github.com/cerner/clara-rules/pull/353/files#r147421772

**Comment by EthanEChristian** [comment]
+1 good catch, updated
